torchvision/prototype/datasets/_builtin/imagenet.py

```diff
@@ -20,176 +19,176 @@
+    }.get(pathlib.Path(data[0]).name)
+
+
+# Although the WordNet IDs (wnids) are unique, the corresponding human-readable categories are not. For example, both
+# 'n02012849' and 'n03126707' are labeled 'crane' while the first means the bird and the latter means the construction
+# equipment.
+WNID_MAP = {
+    "n03126707": "construction crane",
+    "n03710721": "tank suit",
+}
+
+
+def extract_categories_and_wnids(data: Tuple[str, BinaryIO]) -> List[Tuple[str, str]]:
+    synsets = read_mat(data[1], squeeze_me=True)["synsets"]
+    return [
+        (WNID_MAP.get(wnid, category.split(",", 1)[0]), wnid)
+        for _, wnid, category, _, num_children, *_ in synsets
+        # if num_children > 0, we are looking at a superclass that has no direct instance
+        if num_children == 0
+    ]
+
+
+def imagenet_label_to_wnid(imagenet_label: str) -> str:
+    return cast(Tuple[str, ...], WNIDS)[int(imagenet_label) - 1]
+
+
+VAL_TEST_IMAGE_NAME_PATTERN = re.compile(r"ILSVRC2012_(val|test)_(?P<id>\d{8})[.]JPEG")
+
+
+def val_test_image_key(path: pathlib.Path) -> int:
+    return int(VAL_TEST_IMAGE_NAME_PATTERN.match(path.name)["id"])  # type: ignore[index]
+
+
+def prepare_val_data(
+    data: Tuple[Tuple[int, str], Tuple[str, BinaryIO]]
+) -> Tuple[Tuple[Label, str], Tuple[str, BinaryIO]]:
+    label_data, image_data = data
+    _, wnid = label_data
+    label = Label.from_category(WNID_TO_CATEGORY[wnid], categories=CATEGORIES)
+    return (label, wnid), image_data
+
+
+def prepare_sample(
+    data: Tuple[Optional[Tuple[Label, str]], Tuple[str, BinaryIO]],
+) -> Dict[str, Any]:
+    label_data, (path, buffer) = data
+
+    return dict(
+        dict(zip(("label", "wnid"), label_data if label_data else (None, None))),
+        path=path,
+        image=EncodedImage.from_file(buffer),
+    )
+
+
 @register_dataset(NAME)
-class ImageNet(Dataset2):
-    def __init__(self, root: Union[str, pathlib.Path], *, split: str = "train") -> None:
-        self._split = self._verify_str_arg(split, "split", {"train", "val", "test"})
-
-        info = _info()
-        categories, wnids = info["categories"], info["wnids"]
-        self._categories: List[str] = categories
-        self._wnids: List[str] = wnids
-        self._wnid_to_category = dict(zip(wnids, categories))
-
-        super().__init__(root)
-
-    _IMAGES_CHECKSUMS = {
-        "train": "b08200a27a8e34218a0e58fde36b0fe8f73bc377f4acea2d91602057c3ca45bb",
-        "val": "c7e06a6c0baccf06d8dbeb6577d71efff84673a5dbdd50633ab44f8ea0456ae0",
-        "test_v10102019": "9cf7f8249639510f17d3d8a0deb47cd22a435886ba8e29e2b3223e65a4079eb4",
-    }
-
-    def _resources(self) -> List[OnlineResource]:
-        name = "test_v10102019" if self._split == "test" else self._split
-        images = ImageNetResource(
-            file_name=f"ILSVRC2012_img_{name}.tar",
-            sha256=self._IMAGES_CHECKSUMS[name],
+def imagenet(root: Union[str, pathlib.Path], *, split: str = "train", **kwargs: Any) -> TakerDataPipe:
```

Comment:
I don't know yet whether the function should have the same name as the one available in `load()` (i.e. imagenet), or if we should keep the same names as our current dataset to minimize friction (i.e. ImageNet). We can decide that later.

Reply:
The way we'll handle the deprecation / transition is related indeed. That's also something we'll need to decide on. But there is value in conserving the same name, so let's keep an open mind on this.

Reply:
> But there is value in conserving the same name, so let's keep an open mind on this.

The new datasets are BC breaking in so many ways and the consensus from #5040 was that having a compatibility layer is not worth it. Thus, I don't think that the name is the problem here given that they need to do some more work on the entire pipeline.

Reply:
I'm agree that we don't want a compatibility layer.
Still, it's always good to preserve what we can preserve if we can do it for cheap. Again, let's please keep an open mind here.

Reply:
> Still, it's always good to preserve what we can preserve if we can do it for cheap.

I only partially agree. Although it is cheap it brings 2 problems:

1. The old and the new API cannot coexist. Thus, we would have a hard BC break even if users are fine with sticking to the old API for now.
2. Having the same name with different functionality is IMO more confusing than having two separate names one of which will be deprecated in favor of the other.

> Again, let's please keep an open mind here.

:+1: